group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/UniformAssignor.java

```diff
@@ -0,0 +1,288 @@
+         *
+         * @param allTopicIds                   The list of subscribed topic Ids.
+         * @param subscribedTopicDescriber      Utility to fetch the partition count for a given topic.
+         *
+         * @return List of generated {@code TopicIdPartition} for all provided topic Ids.
+         */
+        protected List<TopicIdPartition> allTopicIdPartitions(
+            List<Uuid> allTopicIds,
+            SubscribedTopicDescriber subscribedTopicDescriber
+        ) {
+            List<TopicIdPartition> allTopicIdPartitions = new ArrayList<>();
+            allTopicIds.forEach(topic ->
+                IntStream.range(0, subscribedTopicDescriber.numPartitions((topic)))
+                    .forEach(i -> allTopicIdPartitions.add(new TopicIdPartition(topic, i))
+                )
+            );
+            
+            return allTopicIdPartitions;
+        }
+
+        /**
+         * Represents the rack information of members and partitions along with utility methods
+         * to facilitate rack-aware assignment strategies for a given consumer group.
+         */
+        protected class RackInfo {
+            /**
+             * Map of every member to its rack.
+             */
+            protected final Map<String, String> memberRacks;
+            /**
+             * Map of every partition to a list of its racks.
+             */
+            protected final Map<TopicIdPartition, Set<String>> partitionRacks;
+            /**
+             * Number of members with the same rack as the partition.
+             */
+            private final Map<TopicIdPartition, Integer> numMembersWithSameRackByPartition;
+            protected final boolean useRackStrategy;
+
+            /**
+             * Constructs rack information based on assignment specification and subscribed topics.
+             *
+             * @param assignmentSpec                The current assignment specification.
+             * @param subscribedTopicDescriber      Topic and partition metadata of the subscribed topics.
+             * @param topicIds                      List of topic Ids.
+             */
+            public RackInfo(
+                AssignmentSpec assignmentSpec,
+                SubscribedTopicDescriber subscribedTopicDescriber,
+                Set<Uuid> topicIds
+            ) {
+                Map<String, List<String>> membersByRack = new HashMap<>();
+                assignmentSpec.members().forEach((memberId, assignmentMemberSpec) ->
+                    assignmentMemberSpec.rackId().filter(r -> !r.isEmpty()).ifPresent(
+                        rackId -> membersByRack.computeIfAbsent(rackId, __ -> new ArrayList<>()).add(memberId)
+                    )
+                );
+
+                Set<String> allPartitionRacks;
+                Map<TopicIdPartition, Set<String>> partitionRacks;
+                List<TopicIdPartition> topicIdPartitions = allTopicIdPartitions(new ArrayList<>(topicIds), subscribedTopicDescriber);
```

Comment:
nit: Do we really need to create the ArrayList here?

Reply:
I believe it's easier to understand since allTopicIdPartitions is a method that returns this list which is used later on no?

Reply:
I was referring to `new ArrayList<>(topicIds)`. Is creating a new ArrayList from the Set necessary? allTopicIdPartitions could take a Collection<Uuid> for instance to avoid it.